File: frontend/src/types/production/execution.ts
export type ExecutionStatus = 'in_progress' | 'completed' | 'aborted'

export interface FieldValue {
  field_key: string
  field_label: string
  unit: string | null
  phase: 'start' | 'end'
  value_text: string | null
  value_numeric: number | null
  value_bool: boolean | null
  is_abnormal: boolean
  remark: string | null
}

export interface EquipmentSnapshot {
  equipment_id: string
  equipment_no: string
  equipment_name: string
}

export interface Execution {
  id: string
  batch_id: string
  node_id: string
  node_name: string | null
  execution_seq: number
  status: ExecutionStatus
  owner_id: string | null
  owner_name: string | null
  started_at: string
  started_by_name: string | null
  finished_at: string | null
  finished_by_name: string | null
  is_deviation: boolean
  deviation_reason: string | null
  remark: string | null
  equipments: EquipmentSnapshot[]
  field_values: FieldValue[]
  missing_required_fields?: MissingField[]
}

export interface MissingField {
  field_key: string
  field_label: string
}

export interface FieldValueInput {
  field_key: string
  value: boolean | number | string | null
}

export interface StartExecutionInput {
  node_id: string
  owner_id?: string | null
  owner_name?: string | null
  equipment_ids?: string[]
  field_values?: FieldValueInput[]
  deviation_reason?: string | null
  remark?: string | null
  intermediate_consumptions?: Array<{
    intermediate_type_id: string
    output_id?: string | null  // 精确消耗：选产出批次（与 container_id 二选一）
    container_id?: string | null  // 混装消耗：选容器
    quantity: number
    unit?: string
    remark?: string
  }>
  started_at?: string | null
}

export interface CompleteExecutionInput {
  field_values?: FieldValueInput[]
  remark?: string | null
  intermediate_outputs?: Array<{
    intermediate_type_id: string
    quantity: number
    unit?: string
    intermediate_batch_no?: string
    container_id?: string | null  // 选容器则混装入库
    remark?: string
  }>
  line_id?: string | null
  finished_at?: string | null
}

export interface NodeExecutionListItem {
  id: string
  batch_id: string
  batch_no: string
  execution_seq: number
  status: ExecutionStatus
  owner_name: string | null
  started_at: string
  finished_at: string | null
  is_deviation: boolean
  abnormal_count: number
}

File: backend/app/modules/production/schemas/execution.py
"""节点执行 API 契约。"""

import uuid
from datetime import datetime, timedelta

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.core.time import APP_TZ, now
from app.modules.production.schemas.intermediate import (
    IntermediateConsumptionIn,
    IntermediateOutputIn,
)

# 分钟级时间选择器与服务器时钟之间允许的容差，避免"选当前分钟"被误判为未来
_FUTURE_TOLERANCE = timedelta(minutes=5)


def _normalize_manual_time(v: datetime | None) -> datetime | None:
    """归一化用户手填的工序时间：naive 按厂区时区解释，拒绝未来时间。"""
    if v is None:
        return v
    if v.tzinfo is None:
        v = v.replace(tzinfo=APP_TZ)
    if v > now() + _FUTURE_TOLERANCE:
        raise ValueError("时间不能晚于当前时间")
    return v


class FieldValueIn(BaseModel):
    field_key: str
    value: bool | float | str | None = None


class FieldValueOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    field_key: str
    field_label: str
    unit: str | None
    phase: str
    value_text: str | None
    value_numeric: float | None
    value_bool: bool | None
    is_abnormal: bool
    remark: str | None


class EquipmentSnapshotOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    equipment_id: uuid.UUID
    equipment_no: str
    equipment_name: str


class ExecutionStartIn(BaseModel):
    node_id: uuid.UUID
    owner_id: uuid.UUID | None = None
    owner_name: str | None = Field(default=None, max_length=50)
    equipment_ids: list[uuid.UUID] = []
    field_values: list[FieldValueIn] = []
    deviation_reason: str | None = None
    remark: str | None = None
    intermediate_consumptions: list[IntermediateConsumptionIn] = []
    started_at: datetime | None = None  # 手填开始时间，留空用服务器当前时间

    @field_validator("started_at")
    @classmethod
    def _check_started_at(cls, v: datetime | None) -> datetime | None:
        return _normalize_manual_time(v)


class ExecutionCompleteIn(BaseModel):
    field_values: list[FieldValueIn] = []
    remark: str | None = None
    intermediate_outputs: list[IntermediateOutputIn] = []
    line_id: uuid.UUID | None = None
    finished_at: datetime | None = None  # 手填结束时间，留空用服务器当前时间

    @field_validator("finished_at")
    @classmethod
    def _check_finished_at(cls, v: datetime | None) -> datetime | None:
        return _normalize_manual_time(v)


class ExecutionBackfillIn(BaseModel):
    """工序结束后补录 end 阶段字段值。"""

    field_values: list[FieldValueIn] = []


class MissingFieldOut(BaseModel):
    """已结束工序尚未补录的必填字段（批次结束前须补齐）。"""

    field_key: str
    field_label: str


class ExecutionOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    batch_id: uuid.UUID
    node_id: uuid.UUID
    node_name: str | None = None  # service 组装时填充
    execution_seq: int
    status: str
    owner_id: uuid.UUID | None
    owner_name: str | None
    started_at: datetime
    started_by_name: str | None
    finished_at: datetime | None
    finished_by_name: str | None
    is_deviation: bool
    deviation_reason: str | None
    remark: str | None
    equipments: list[EquipmentSnapshotOut] = []
    field_values: list[FieldValueOut] = []
    missing_required_fields: list[MissingFieldOut] = []  # service 组装：已结束工序缺填的必填字段


class NodeExecutionListItem(BaseModel):
    """工序视角的执行记录行（跨批次）。"""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    batch_id: uuid.UUID
    batch_no: str
    execution_seq: int
    status: str
    owner_name: str | None
    started_at: datetime
    finished_at: datetime | None
    is_deviation: bool
    abnormal_count: int
